src/vue/directives/tooltip.js

```diff
@@ -0,0 +1,387 @@
+      case DIRECTIONS.right:
+        positionStyle = {
+          left: twRect.width + MARGIN_SIZE,
+        }
+        break
+      case DIRECTIONS.left:
+        positionStyle = {
+          left: 0 - twRect.width - MARGIN_SIZE,
+        }
+        break
+      case DIRECTIONS.bottom:
+      default:
+        positionStyle = {
+          top: 0 - targetHeight - MARGIN_SIZE,
+        }
+        break
+    }
+    return positionStyle
+  }
+
+  function renderTooltip (e, binding) {
+    target = e.target ? e.target : e
+    targetRect = target.getBoundingClientRect()
+    targetHeight = targetRect.height
+    position = binding.value.position
+    currentPosition = position
+    bgColor = binding.value.bgColor || COLOR_BLACK
+    textColor = binding.value.textColor || COLOR_WHITE
+    fontSize = binding.value.fontSize || FONT_SIZE
+    width = binding.value.width || WIDTH
+    height = binding.value.height || HEIGHT
+    tooltipWrapper = document.createElement('div')
+    tooltipText = document.createElement('div')
+    tooltipArrow = document.createElement('div')
+    tooltipWrapper.appendChild(tooltipText)
+    tooltipWrapper.appendChild(tooltipArrow)
+    target.appendChild(tooltipWrapper)
+    target.classList.add('tooltip-container')
+
+    tooltipWrapper.className = 'tooltip-wrapper'
+    tooltipWrapper.style.position = 'absolute'
+    tooltipWrapper.style.width = width
+    tooltipWrapper.style.height = height
+
+    tooltipText.className = 'tooltip-wrapper-text'
+    tooltipText.innerText = globalize(binding.value.text)
+    tooltipText.style.fontSize = fontSize
+    tooltipText.style.color = textColor
+
+    twRect = tooltipWrapper.getBoundingClientRect()
+    const positionStyle = getPositionStyle(twRect)
+
+    tooltipWrapper.style.pointerEvents = 'none'
+    tooltipWrapper.style.top = positionStyle.top + 'px'
+    tooltipWrapper.style.right = positionStyle.right + 'px'
+    tooltipWrapper.style.bottom = positionStyle.bottom + 'px'
+    tooltipWrapper.style.left = positionStyle.left + 'px'
+    tooltipWrapper.style.backgroundColor = bgColor
+    tooltipWrapper.style.display = 'flex'
+    tooltipWrapper.style.justifyContent = 'center'
+    tooltipWrapper.style.alignItems = 'center'
+    tooltipWrapper.style.padding = '8px'
+    tooltipWrapper.style.transition = 'transform 350ms ease-in-out'
+    tooltipWrapper.style.opacity = '0'
+    tooltipWrapper.style.zIndex = '120'
+    tooltipWrapper.style.borderRadius = '4px'
+
+    if (target.style.position !== 'relative') {
+      target.style.position = 'relative'
+    }
+
+    alignTooltip(target)
+    opacity = 0
+    animateFadeEffect()
+
+    tooltipWrapper.addEventListener('transitionend', function initRenderArrow () {
+      renderArrow(target)
+      tooltipWrapper.removeEventListener('transitionend', initRenderArrow)
+    })
+  }
+
+  function animateFadeEffect () {
+    if (tooltipWrapper.style.opacity < 1) {
+      opacity += 0.1
+      setTimeout(animateFadeEffect, 100)
```

Comment:
Why not just use the transition for the opacity?

Reply:
By design, got approve that its acceptable

Reply:
But we've discussed that it's will make the transition with 9fps, but not 60fps.
Also, here you're set opacity into 1.2 value, but not 1. It will work fine in popular browsers right now, but it may occur some problems somewhere in the future or right now too because you don't know how it works on some Samsung browser or etc (it does not matter where exactly, its just example)

Reply:
Refactored